### PR TITLE
enhance(entrypoint): separate stable entrypoint from sw client entry

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description

1. origin image schema:
```
  base
     |  \ 
     |    \ 
     |      cuda
     |    /
     |  /
  starwhale(preinstalled the special sw-cli version and entrypoint)
```
2. current image schema:
```
  base(stable-entrypoint)
      |
      |
   cuda
```
The stable-entrypoint will config environment(install sw-cli) when container is created.  

## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

**Review by jialeicui — APPROVED** [review]
LGTM

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/DatasetService.java [hunk 1]
```diff
@@ -464,14 +464,24 @@ public void build(CreateBuildRecordRequest request) {
...
             var user = userService.currentUserDetail();
-            DockerImage image;
-            var runtimeProperties = systemSettingService.getRunTimeProperties();
-            if (runtimeProperties != null && runtimeProperties.getDatasetBuild() != null
-                    && StringUtils.hasText(runtimeProperties.getDatasetBuild().getImage())) {
-                image = new DockerImage(systemSettingService.getRunTimeProperties().getDatasetBuild().getImage());
-            } else {
-                image = new DockerImage("docker-registry.starwhale.cn/star-whale/starwhale:latest");
+            DockerImage image = new DockerImage("docker-registry.starwhale.cn/star-whale/base:latest");
+            String swVersion = "0.5.6";
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/DatasetService.java:468]: is it possible to make this configurable in system setting

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/ModelServingService.java [hunk 2]
```diff
@@ -286,6 +286,8 @@ private void deploy(
         var md = modelMapper.find(model.getModelId());
 
         var envs = new HashMap<String, String>();
+        envs.put("SW_RUNTIME_PYTHON_VERSION", runtime.getVersionMetaObj().getEnvironment().getPython());
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/ModelServingService.java:289]: as python version and sw version are mandatory， how about populate them directlly from runtime such as:
`runtime.getSwVersion()` `runtime.getPythonVersion()`

File: client/scripts/sw-docker-entrypoint
Comment (anda-ren) [inline client/scripts/sw-docker-entrypoint:1]: remove code server installation ?

File: docker/Dockerfile.base [hunk 3]
```diff
@@ -52,3 +59,12 @@ RUN python3.7 -m pip install --upgrade virtualenv>=20.14.0 pip>=22.1.0 \
     && curl -sS https://bootstrap.pypa.io/get-pip.py | python3.11 \
     && python3.11 -m pip install --upgrade virtualenv>=20.14.0 pip>=22.1.0 \
     && python3 -m pip install yq && python3 -m pip cache purge
+
+# code-server
+RUN curl -fsSL --compressed  https://code-server.dev/install.sh | sh
+
+# stable entrypoint
+WORKDIR /opt/starwhale.bin
+COPY external/starwhale-entrypoint.sh starwhale-entrypoint
```
Comment (tianweidut) [inline docker/Dockerfile.base:68]: use another name for `starwhale-entrypoint.sh` or `sw-docker-entrypoint.sh`?

`sw-docker-entrypoint` and `starwhale-entrypoint` maybe lead to a bit confused.

File: docker/external/starwhale-entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    echo "-->[Preparing] Set pip config."
+    set_pip_config
+
+    echo "-->[Preparing] Use python:${PYTHON_VERSION}."
+    set_python_alter
+    set_pip_cache
+
+    echo "-->[Preparing] Install starwhale:${SW_VERSION}."
+    # install starwhale for current python
+    python3 -m pip install "starwhale==${SW_VERSION}" || exit 1
```
Comment (tianweidut) [inline docker/external/starwhale-entrypoint.sh:76]: `SW_VERSION` is a required value for the script.
If not set SW_VERSION, can we exit 1 directly or `pip install starwhale` is ok?

File: docker/external/starwhale-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+case "$1" in
+    set_environment)
+        set_py_and_sw
+        ;;
+    *)
+        if [ "${RUNTIME_RESTORED}" != "1" ]; then
+          set_py_and_sw
+        fi
+        if [ -z "${USE_CUSTOM_CMD}" ]; then
```
Comment (tianweidut) [inline docker/external/starwhale-entrypoint.sh:87]: what's the meaning of `USE_CUSTOM_CMD`?

File: docker/external/starwhale-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    echo "-->[Preparing] Use python:${PYTHON_VERSION}."
+    set_python_alter
+    set_pip_cache
+
+    echo "-->[Preparing] Install starwhale:${SW_VERSION}."
+    # install starwhale for current python
+    python3 -m pip install "starwhale==${SW_VERSION}" || exit 1
+}
+
+case "$1" in
```
Comment (tianweidut) [inline docker/external/starwhale-entrypoint.sh:79]: provide a help output for the `starwhale-entrypoint.sh`?

for example, the required envs, and the flag?

File: docker/external/starwhale-entrypoint.sh (hunk 4 above)
Comment (tianweidut) [inline docker/external/starwhale-entrypoint.sh:76]: how to use the starwhale from the GitHub git version?

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/DatasetService.java (hunk 1 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/DatasetService.java]: done

File: client/scripts/sw-docker-entrypoint
Comment (goldenxinxing) [reply client/scripts/sw-docker-entrypoint:1]: done

File: docker/Dockerfile.base (hunk 3 above)
Comment (goldenxinxing) [reply docker/Dockerfile.base]: done

File: docker/external/starwhale-entrypoint.sh (hunk 4 above)
Comment (goldenxinxing) [reply docker/external/starwhale-entrypoint.sh]: fixed

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/ModelServingService.java (hunk 2 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/ModelServingService.java]: done

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**